File: src/cli/taskRunner.ts
import { Coordinator } from '../agents/coordinator.ts';
import { classifyTask, conversationReply } from '../agents/fastRouter.ts';
import { RoleModelRouter } from '../agents/roleRouter.ts';
import type { CoordinatorRunResult } from '../agents/types.ts';
import { TaskCancelledError } from '../agent/loop.ts';
import { reportRunUsage } from '../auth/reporting.ts';
import { ProjectIndexService } from '../project/index.ts';
import { RemoteProvider } from '../providers/remote.ts';
import type { Runtime } from '../runtime.ts';
import { c } from './colors.ts';
import { ProgressRenderer } from './ui/progress.ts';
import { renderError, renderTaskResult } from './ui/results.ts';
import { symbols } from './ui/theme.ts';
import { isVerbose } from './verbose.ts';
import { sessionRolloverNote, sessionStatusLine } from './freePlan.ts';

export interface TaskRunOptions {
  /** Await the usage report (one-shot) vs fire-and-forget (REPL). */
  awaitUsageReport?: boolean;
  /** Show raw diagnostics (plan, agent details, more output). */
  verbose?: boolean;
  /** Abort signal: Ctrl+C during a task cancels the run safely. */
  signal?: AbortSignal;
}

/** One shared, maintained project index per runtime (survives across tasks). */
const indexByRuntime = new WeakMap<object, ProjectIndexService>();

/**
 * Run one user task through the GRACE coordinator.
 *
 * Renders concise structured progress (`· Planning… → Project Scout ✓ …`)
 * without exposing any agent chain-of-thought, then prints composed result
 * sections (Done / Files changed / Validation / Provider / Time / follow-ups),
 * the GRACE FREE quota line and usage reporting.
 */
export async function runTask(runtime: Runtime, input: string, opts: TaskRunOptions = {}): Promise<number> {
  // Greetings never need a provider, an index or any agent — answer instantly.
  const route = classifyTask(input).route;
  if (route === 'conversation') {
    console.log(conversationReply(input));
    return 0;
  }

  if (!runtime.provider) {
    console.log(renderError(runtime.providerError ?? 'No AI provider configured.'));
    return 1;
  }

  let index = indexByRuntime.get(runtime);
  if (!index) {
    index = new ProjectIndexService(runtime.root);
    indexByRuntime.set(runtime, index);
  }

  console.log('');
  const startedAt = Date.now();
  // The primary agent uses the user's configured provider/model directly — one
  // visible provider, no per-role switching. Optional specialists (complex
  // plans) still resolve their own route through the Model Router.
  const roleRouter = new RoleModelRouter(runtime);
  const served = runtime.provider instanceof RemoteProvider ? (runtime.provider.serverProvider ?? RemoteProvider.sharedServerProvider()) : null;
  const verbose = opts.verbose ?? isVerbose();
  const progress = new ProgressRenderer({
    verbose,
    // Provider/model are debug output — normal users see them in /status.
    providerLabel: verbose ? (served?.label ?? runtime.provider.label) : undefined,
    model: verbose ? runtime.provider.getModel().id : undefined,
  });
  const coordinator = new Coordinator({
    runtime,
    projectIndex: index,
    onEvent: (e) => progress.event(e),
    providerFactory: (role, spec) => (role === 'editor' ? runtime.provider : roleRouter.providerFor(role, spec)),
    plannerProvider: roleRouter.plannerProvider(),
    signal: opts.signal,
  });

  let result: CoordinatorRunResult;
  try {
    result = await coordinator.run(input);
  } catch (err) {
    progress.end();
    if (err instanceof TaskCancelledError) {
      console.log('');
      console.log(c.dim('Cancelled.'));
      return 0;
    }
    console.log(renderError('Task failed', (err as Error).message ?? String(err)));
    return 1;
  }
  progress.end();

  const executionTimeMs = Date.now() - startedAt;
  console.log('');

  // The editor may have changed files — make sure the next task sees a fresh index.
  if (result.changedFiles.length > 0) index.invalidate();

  // A greeting is answered directly — no result circus.
  if (result.route === 'conversation') {
    console.log(result.finalAnswer);
    return 0;
  }

  console.log(renderTaskResult({ result, runtime, executionTimeMs, verbose }));

  // GRACE FREE: daily session quota from the server's last response. Role
  // routing creates one RemoteProvider per agent tier, so the freshest state
  // is the shared view across all instances of this run.
  if (runtime.provider instanceof RemoteProvider) {
    const last = RemoteProvider.sharedSession();
    if (last) {
      const line = sessionStatusLine(last);
      if (line) console.log(line);
      if (last.startedNew) console.log(sessionRolloverNote(last));
    }
  }

  const report = () =>
    reportRunUsage({
      prompt: input,
      model: runtime.provider!.getModel().id,
      projectType: runtime.project.type,
      iterations: result.iterations,
      toolCalls: result.toolCalls,
      usage: result.usage,
      executionTimeMs,
    }).then((outcome) => {
      if (outcome === 'failed') {
        console.log(`${symbols().bullet} usage report failed (backend offline) — run continued locally.`);
      }
    });

  if (opts.awaitUsageReport) {
    await report();
  } else {
    void report();
  }

  return 0;
}

File: grace/cli/task_runner.py
"""Task execution through the coordinator (port of src/cli/taskRunner.ts).

Renders concise structured progress without exposing any agent chain-of-
thought, then prints composed result sections (Done / Files changed /
Validation / Provider / Time / follow-ups), the GRACE FREE quota line and
usage reporting.
"""

import time

from grace.agent.loop import TaskCancelledError
from grace.agents.coordinator import Coordinator
from grace.agents.fast_router import classify_task, conversation_reply
from grace.agents.role_router import RoleModelRouter
from grace.auth.reporting import report_run_usage
from grace.auth.session import load_session, session_expired
from grace.cli.free_plan import session_rollover_note, session_status_line
from grace.cli.ui.progress import ProgressRenderer
from grace.cli.ui.results import render_error, render_task_result
from grace.cli.ui.theme import symbols
from grace.colors import c
from grace.project.index import ProjectIndexService
from grace.providers.remote import RemoteProvider
from grace.verbose import is_verbose

# One shared, maintained project index per runtime (survives across tasks).
_index_by_runtime: dict[int, ProjectIndexService] = {}


def run_task(runtime, input_text: str, opts: dict | None = None) -> int:
    """Run one user task through the GRACE coordinator. Returns an exit code."""
    opts = opts or {}
    # Greetings never need a provider, an index or any agent — answer instantly.
    route = classify_task(input_text)
    if route == "conversation":
        print(conversation_reply(input_text))
        return 0

    if not runtime.provider:
        print(render_error(runtime.provider_error or "No AI provider configured."))
        return 1

    index = _index_by_runtime.get(id(runtime))
    if index is None:
        index = ProjectIndexService(runtime.root)
        _index_by_runtime[id(runtime)] = index

    print("")
    started_at = time.monotonic() * 1000
    role_router = RoleModelRouter(runtime)
    served = runtime.provider.server_provider if isinstance(runtime.provider, RemoteProvider) else None
    verbose = opts.get("verbose") if opts.get("verbose") is not None else is_verbose()
    progress = ProgressRenderer({
        "verbose": verbose,
        "providerLabel": (served or {}).get("label") or runtime.provider.label if verbose else None,
        "model": runtime.provider.get_model().id if verbose else None,
    })
    coordinator = Coordinator({
        "runtime": runtime,
        "projectIndex": index,
        "onEvent": progress.event,
        "providerFactory": (lambda role, spec: runtime.provider if role == "editor" else role_router.provider_for(role, spec)),
        "plannerProvider": role_router.planner_provider(),
        "signal": opts.get("signal"),
    })

    try:
        result = coordinator.run(input_text)
    except TaskCancelledError:
        progress.end()
        print("")
        print(c.dim("Cancelled."))
        return 0
    except Exception as err:
        progress.end()
        print(render_error("Task failed", str(err)))
        return 1
    progress.end()

    execution_time_ms = time.monotonic() * 1000 - started_at
    print("")

    # The editor may have changed files — make sure the next task sees a fresh index.
    if result.changedFiles:
        index.invalidate()

    if result.route == "conversation":
        print(result.finalAnswer)
        return 0

    print(render_task_result({"result": result, "runtime": runtime, "executionTimeMs": execution_time_ms, "verbose": verbose}))

    # GRACE FREE: daily session quota from the server's latest response.
    stored = load_session()
    if stored and not session_expired(stored):
        last = RemoteProvider.shared_session()
        if last:
            line = session_status_line(last)
            if line:
                print(line)
            if last.get("startedNew"):
                print(session_rollover_note(last))

    def report() -> str:
        outcome = report_run_usage({
            "prompt": input_text,
            "model": runtime.provider.get_model().id,
            "projectType": runtime.project.type,
            "iterations": result.iterations,
            "toolCalls": result.toolCalls,
            "usage": result.usage,
            "executionTimeMs": execution_time_ms,
        })
        if outcome == "failed":
            print(f"{symbols()['bullet']} usage report failed (backend offline) — run continued locally.")
        return outcome

    if opts.get("awaitUsageReport"):
        report()

    return 0
